### ISSUE TITLE
Bug with level trigger on windows

### ISSUE THREAD

**Issue by GGist** [carllerche/mio#336]
I have a use case for mio where I need to read some bytes from a stream (may not be all of the bytes available in the stream), then write some bytes to the stream, and finally read the last bytes from the stream. Currently I am registering TcpStreams in level triggered mode to achieve this however, on windows, it seems like the read event is not being generated even when I have not read all of the bytes in the stream.

Here is a stripped down example for what I am wanting to achieve, on windows, the message ```Going into Finished state...``` is never printed out meaning we never read the last bytes in the stream in the ```SimpleState::SecondRead``` state. However, I tested on Ubuntu and it does get printed out as the last bytes are read.

```
extern crate bytes;
extern crate mio;

use std::net::{self, Ipv4Addr, SocketAddr, SocketAddrV4};
use std::io::{Write, Read};
use std::thread::{self};

use bytes::{Take, ByteBuf, MutByteBuf, MutBuf, Buf};
use mio::{EventLoop, Token, EventSet, PollOpt, Handler, TryRead, TryWrite};
use mio::tcp::{self, TcpListener};

enum SimpleState {
    FirstRead(Take<MutByteBuf>),
    Write(ByteBuf),
    SecondRead(Take<MutByteBuf>)
}

enum NextState {
    Write,
    SecondRead,
    Finished
}

struct SimpleClient {
    conn:  tcp::TcpStream,
    state: SimpleState
}

struct SimpleHandler {
    listener:   (Token, TcpListener),
    opt_client: Option<(Token, SimpleClient)>,
}

impl Handler for SimpleHandler {
    type Timeout = ();
    type Message = ();
    
    fn ready(&mut self, event_loop: &mut EventLoop<SimpleHandler>, token: Token, events: EventSet) {
        if events.is_error() | events.is_hup() {
            panic!("Received error or hup event");
        }
        
        if self.listener.0 == token {
            // Store our single connection
            let stream = self.listener.1.accept().unwrap().unwrap().0;
            let simple_client = SimpleClient{ conn: stream, state: SimpleState::FirstRead(Take::new(ByteBuf::mut_with_capacity(20), 20)) };
            let simple_token = Token(2);
            
            event_loop.register(&simple_client.conn, simple_token, EventSet::readable(), PollOpt::level() | PollOpt::oneshot()).unwrap();
            println!("Going into FirstRead state...");
            self.opt_client = Some((simple_token, simple_client));
        } else if self.opt_client.as_ref().unwrap().0 == token {
            // Process an event for our single connection
            let ref mut client = self.opt_client.as_mut().unwrap().1;
            
            // Consume the bytes for the event we subscribed to
            let (opt_next_state, event_set) = match client.state {
                SimpleState::FirstRead(ref mut buf) => {
                    let has_remaining = client.conn.try_read_buf(buf).map(|_| buf.has_remaining()).unwrap();
                    
                    if !has_remaining {
                        (Some(NextState::Write), EventSet::writable())
                    } else {
                        (None, EventSet::readable())
                    }
                },
                SimpleState::Write(ref mut buf) => {
                    let has_remaining = client.conn.try_write_buf(buf).map(|_| buf.has_remaining()).unwrap();
                    
                    if !has_remaining {
                        (Some(NextState::SecondRead), EventSet::readable())
                    } else {
                        (None, EventSet::writable())
                    }
                },
                SimpleState::SecondRead(ref mut buf) => {
                    let has_remaining = client.conn.try_read_buf(buf).map(|_| buf.has_remaining()).unwrap();
                    
                    if !has_remaining {
                        (Some(NextState::Finished), EventSet::none())
                    } else {
                        (None, EventSet::readable())
                    }
                }
            };
            
            // Check if we need to advance a state or not
            match opt_next_state {
                Some(NextState::Write) => {
                    println!("Going into Write state...");
                    client.state = SimpleState::Write(ByteBuf::from_slice(&[0u8; 20][..]));
                },
                Some(NextState::SecondRead) => {
                    println!("Going into SecondRead state...");
                    client.state = SimpleState::SecondRead(Take::new(ByteBuf::mut_with_capacity(20), 20));
                },
                Some(NextState::Finished) => println!("Going into Finished state..."),
                None => ()
            };
            
            // Reregister ourselves
            event_loop.reregister(&client.conn, token, event_set, PollOpt::level() | PollOpt::oneshot());
        }
    }
}

fn main() {
    let listen_ip = Ipv4Addr::new(127, 0, 0, 1);
    let listen_addr = SocketAddr::V4(SocketAddrV4::new(listen_ip, 8876));
    
    // Thread for the event loop
    thread::spawn(move || {
        let mut event_loop = EventLoop::new().unwrap();
        
        let listener = TcpListener::bind(&listen_addr).unwrap();
        let mut simple_handler = SimpleHandler{ listener: (Token(1), listener), opt_client: None };
        
        event_loop.register(&simple_handler.listener.1, simple_handler.listener.0, EventSet::readable(), PollOpt::edge()).unwrap();
        
        event_loop.run(&mut simple_handler).unwrap();
    });
    // Wait for event loop to start
    thread::sleep_ms(1000);
    
    let mut stream = net::TcpStream::connect("127.0.0.1:8876").unwrap();
    let mut recv = [0u8; 20];
    
    // Write 40 bytes (other end of stream will only read 20 at first, then will write 20 bytes, then SHOULD read the last 20 bytes)
    stream.write_all(&[0u8; 40]).unwrap();
    
    // Read 20 bytes that the other end wrote
    &stream.read_exact(&mut recv).unwrap();
    
    // Sleep for a few ms to let the connection finish
    thread::sleep_ms(5000);
}
```

Tested using mio 0.5 with both of the following configurations:

OS: Windows 7 64-bit
rustc 1.7.0-nightly (7ce713961 2015-12-11)
binary: rustc
commit-hash: 7ce713961cb7c36bb66db70a276973eb78cf9e6e
commit-date: 2015-12-11
host: x86_64-pc-windows-gnu
release: 1.7.0-nightly

OS: Ubuntu 14.04 64-bit
rustc 1.7.0-nightly (3150dddb0 2015-12-25)
binary: rustc
commit-hash: 3150dddb0f2f6c84f2355db324794eecab6a3299
commit-date: 2015-12-25
host: x86_64-unknown-linux-gnu
release: 1.7.0-nightly

**Comment by carllerche** [comment]
Thanks for the report. I will look into this.

**Comment by carllerche** [comment]
Finally looking at this. So sorry for the delay.

First, I find it odd that you are using level triggered w/ one shot. It seems like it would be better to just use edge triggered + oneshot in this case. I believe that this will satisfy your needs.

I'm going to most likely require oneshot to be used w/ edge triggered for now at least.

However, simply changing to edge triggered doesn't seem to resolve the issue, so I will need to dig in more.

**Comment by GGist** [comment]
No problem at all.

So I haven't used APIs like ```epoll``` or ```kqueue``` before coming to ```mio``` so my understanding of ```PollOpt``` may not be correct. However in this case, I am using ```PollOpt::oneshot()``` because when I am in a state transition, my ```EventSet``` will change from ```EventSet::readable()``` to ```EventSet::writable()``` or vice versa and I want to re-register with the new ```EventSet```.

I don't believe ```PollOpt::edge()``` would satisfy my needs because as I understand it, an edge triggered event is only generated when data is received (or when the kernel buffer is ready for writing). For example, if a client sends me 40 bytes an event would be generated and in my case I only want to read 20 bytes (not all of the bytes), then I want to write 20 bytes to them and afterwards read 20 bytes from them. Since I already handled the edge triggered event when reading the first 20 bytes, I would never be able to read the last 20 bytes that I didn't read initially.

So ```PollOpt::level()``` would help me here in that after I write my 20 bytes I will still get an event generated that data is ready to be read since level triggering would constantly annoy me until all of the data was read.

Let me know if any of my assumptions are incorrect. Also, if level triggering combined with oneshot has some technical issues on windows that need to be resolved, I think just having a working level triggering available would suffice.

**Comment by carllerche** [comment]
You can use edge + oneshot here because if you receive a read notification, read part of the data, re-reregister the socket w/ edge + oneshot, you will get a notification immediately.

I have a fix in progress here: https://github.com/carllerche/mio/pull/347 If you can test it by setting mio in your Cargo.toml to pull from git on the win-fix-oneshot branch, it should validate if the fix is correct.

Also, here is the test, you can see that I am using edge + oneshot and doing partial reads :)

https://github.com/carllerche/mio/blob/win-fix-oneshot/test/test_oneshot.rs

Thanks for the report and the detailed repro... it helped a lot to debug.

**Comment by GGist** [comment]
Terrific, I didn't realize edge triggered events are checked again upon re-registration.

I switched my level triggers over to edge triggers and added the ```win-fix-oneshot``` branch in my ```Cargo.toml``` and my tests are now passing and the original example is working as intended as well.

Thanks a lot!

**Comment by carllerche** [comment]
Great! I would definitely wait until the fix lands on master before pushing anything into production though!

**Comment by GGist** [comment]
Definitely, no rush.

**Comment by carllerche** [comment]
Ok, so this took longer than I had hoped, but the fix is merged on master (#356). Thanks!
